source/common/http/http2/codec_impl.cc

```diff
@@ -900,65 +905,83 @@ void ConnectionImpl::sendPendingFrames() {
+    const auto it = std::find_if(settings.cbegin(), settings.cend(),
+                                 [&entry](const nghttp2_settings_entry& existing) {
+                                   return entry.settings_id == existing.settings_id;
+                                 });
+    if (it != settings.end()) {
+      return false;
+    }
+    settings.push_back(entry);
+    return true;
+  };
+
+  // Universally disable receiving push promise frames as we don't currently support
+  // them. nghttp2 will fail the connection if the other side still sends them.
+  // TODO(mattklein123): Remove this when we correctly proxy push promise.
+  // NOTE: This is a special case with respect to custom parameter overrides in that server push is
+  // not supported and therefore not end user configurable.
   if (disable_push) {
-    // Universally disable receiving push promise frames as we don't currently support them. nghttp2
-    // will fail the connection if the other side still sends them.
-    // TODO(mattklein123): Remove this when we correctly proxy push promise.
-    iv.push_back({NGHTTP2_SETTINGS_ENABLE_PUSH, 0});
-  }
-
-  if (!iv.empty()) {
-    int rc = nghttp2_submit_settings(session_, NGHTTP2_FLAG_NONE, &iv[0], iv.size());
+    settings.push_back(
+        {static_cast<int32_t>(NGHTTP2_SETTINGS_ENABLE_PUSH), disable_push ? 0U : 1U});
+  }
+
+  for (const auto it : http2_options.custom_settings_parameters()) {
+    // Validate that named parameters (with the exception of `allow_connect`) are not being
+    // overridden by custom parameters.
+    ASSERT((absl::c_none_of<std::array<uint16_t, 4>>(
+        {NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, NGHTTP2_SETTINGS_ENABLE_PUSH,
+         NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE},
+        [it](const uint16_t& entry) { return entry == it.identifier().value(); })));
+    ASSERT(it.identifier().value() <= std::numeric_limits<uint16_t>::max());
+
+    const bool result =
+        insertParameter({static_cast<int32_t>(it.identifier().value()), it.value().value()});
+    if (!result) {
+      ENVOY_CONN_LOG(debug, "duplicate custom settings parameter with id {:#x}, value {}",
+                     connection_, it.identifier().value(), it.value().value());
+      continue;
+    }
+    ENVOY_CONN_LOG(debug, "adding custom settings parameter with id {:#x} to {}", connection_,
+                   it.identifier().value(), it.value().value());
+  }
+
+  // Insert named parameters.
+  settings.insert(
+      settings.end(),
+      {{NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, http2_options.hpack_table_size().value()},
+       {NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_options.max_concurrent_streams().value()},
+       {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_options.initial_stream_window_size().value()}});
+  const bool result =
+      insertParameter({NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL, http2_options.allow_connect()});
+  if (!result) {
+    ENVOY_CONN_LOG(warn,
```

Comment:
This could log at high rate in the data path, right? I think we need to do this at debug level like above? 

To clarify, are both this case and the above case all to handle the allow_connect case? Can't we detect/warn about all of this during config validation? I'm unclear why we need to do this again here?

Reply:
Thanks, this was an oversight. Changed it to a debug log. The ideal solution would be to have a  `LOG_EVERY_N_SEC()` macro that would allow me to log at warn level but rate limit it to avoid the perf hit and spamming the log (note to self: implement this).

The issue is that `allow_connect` can not be validated during config load since its presence can not be checked. To enable forward API compatibility, I only want to fail config load when both the named field and a user defined parameter with the same identifier are set.

Reply:
> The ideal solution would be to have a LOG_EVERY_N_SEC() macro that would allow me to log at warn level but rate limit it to avoid the perf hit and spamming the log (note to self: implement this).

FWIW this comes up about once per month, but no one has implemented it yet. :) cc @jmarantz 

> The issue is that allow_connect can not be validated during config load since its presence can not be checked. To enable forward API compatibility, I only want to fail config load when both the named field and a user defined parameter with the same identifier are set.

I'm confused about this. Why can't the presence be checked? Isn't it also config?

Reply:
`allow_connect` has a primitive (`bool`) type, so it's not possible to differentiate between the default value and "set with the same value as the default". We need to use a wrapped type (`google.protobuf.BoolValue`) to be able to check presence.

Reply:
Ah I see, sorry. Can you leave more of a comment trail around that in the data plane code also about the proto type? That would have helped me.

Reply:
Will do.

Reply:
Right but if we can't check for presence we can still fail on conflicting values no?
It's annoying to have it set in 2 places.  It's buggy (IMO) if those two places disagree.

Reply:
+1 if we can do the disagreement check. If so IMO we can drop the debug logs in the data path?

Reply:
SG; I'll make the change to fail config on conflicting values.

Reply:
The validation logic has now been simplified. PTAL.